llvm/lib/Transforms/Utils/SimplifyCFG.cpp

```diff
@@ -1742,15 +1767,6 @@ bool SimplifyCFGOpt::hoistCommonCodeFromSuccessors(BasicBlock *BB,
       continue;
     }
 
-    SmallVector<Instruction *, 8> OtherInsts;
-    if (SameLevelHoist) { 
```

Comment:
Why did you remove `SameLevelHoist`?
I think it makes sense. You can build the hash map when you can't hoist at the same level. This can reduce the size of it.

Reply:
I'm still doing same level hoisting but those instructions are in the hash map as well. If I want to avoid putting same level instructions in the hash map, the logic for filling the map will become more complicated and needs iterating BBs more than once which brings overhead. Take this case for example: 
```
    BB1           BB2
    I0             I2
    I1             I1
    I2             I3
```
Let's say instructions with same numbers are identical.

 While checking `I0` from `BB1`, if I want to check the same level instruction of `BB2` (which is `I2` here), after finding that `I0` and `I2` are not identical I need to put `I2` in the map (hoping that I will find an identical instruction from BB1 later) but that would also increase map size if no identical instruction is found. The other option is not putting `I2` in the map at this point, but wait until we are done iterating BB1, then re-iterate BB2. In this way when we reach `I2` in BB2, we have already added its identical instruction from BB1 so it can be added to the map. But this is not desirable as we want to iterate each BB only once.

Reply:
I think there is a solution here that splits the hoist common code into three steps (function or class?).

1. The current simple matching method.
2. Use hash map for matching.
3. Hoist the termination instruction.

Also, I think with the hash map, we can simplify the first step of the logic by removing the skip part and going straight to the second step once we encounter a difference.

Could you create a PR first to prepare for this change? Perhaps, we just need to split the current code into two (1 and 3) steps in order to insert the hash map code. :)

Reply:
You're right, it makes sense to split the code into 3 steps to address those cases independently. Just as a bug fix, can we merge the current code? I think it covers the problem that was mentioned in the original issue.  I can work on your proposal in a separate PR a bit later :)

Reply:
For me personally, optimization is not about fixing bugs. But others may have a different idea.
@nikic @dtcxzyw

Reply:
What is your idea @nikic ?
It's reasonable to make those changes but I assume it will take a bit time and won't be able to do it immediately (probably I will face with some corner cases and need to resolve them). This PR has been open for a long time and we have good results in terms of code size. What do you suggest?

Reply:
Gently ping. @nikic @dtcxzyw 

(Maybe we should pay more consideration to first-time contributors.)

Reply:
ping @DianQK

Reply:
I'm just finding that the code here is getting more complicated. Of course, a big part of the problem here is my previous commits :3.
There is a possibility of ambiguity in this change, and I can verify my thoughts in subsequent PRs.

(Sorry, my github notifications are starting to get hard to sift through the important parts.)